pkg/app/validate_app.go

```diff
@@ -118,7 +126,55 @@ func (v *Validator) Validate(request *admissionv1.AdmissionRequest) (bool, error
 		return false, microerror.Mask(err)
 	}
 
+	err = v.emitEvents(ctx, request, app)
+	if err != nil {
+		v.logger.Errorf(ctx, err, "app %#q has failed to emit events", app.Name)
+	}
+
 	v.logger.Debugf(ctx, "admitted app %#q in namespace %#q", app.Name, app.Namespace)
 
 	return appAllowed, nil
 }
+
+func (v *Validator) emitEvents(ctx context.Context, request *admissionv1.AdmissionRequest, app v1alpha1.App) error {
+	if request.Operation != admissionv1.Update {
+		// no-op when it's not update
+		return nil
+	}
+
+	var oldApp v1alpha1.App
+
+	if _, _, err := validator.Deserializer.Decode(request.OldObject.Raw, nil, &oldApp); err != nil {
+		return microerror.Maskf(parsingFailedError, "unable to parse app: %#v", err)
+	}
+
+	compareFunc := map[string]func(v1alpha1.App) string{
```

Comment:
I think having a list of compare functions like below is much tidy.

Reply:
Yes makes sense but also could be useful in other places.

What about moving to the key package in the app library? Since it's using lots of key funcs anyway.

Reply:
Also offtopic but the logging with `namespace/name` is nice.

I think @fiunchinho did that in apptest as well. We should improve the logging in app- and chart-operator too.

Reply:
@rossf7 Do you mean moving `compareFunc` into the app library? But here I would like to show users about events that relevant to the user interface when we hide other changes (e.g. annotation, kubeconfig changes). 

Both app-operator and chart-operator are using `cmp.diff` from the desired chart CRs, so there are not many places we could use this list other than app-admission-controller. So I'm inclined to stay this function here only, WDYT?

Reply:
OK sure if we're not going to reuse it then it's more readable to have it here IMO. 👍 

But the ordering is weird here. Can we have this in alpha order?